server/service/mdm.go

```diff
@@ -2109,3 +2112,109 @@ func (svc *Service) ResendHostMDMProfile(ctx context.Context, hostID uint, profi
+
+func getMDMAppleCSREndpoint(ctx context.Context, request interface{}, svc fleet.Service) (errorer, error) {
+	signedCSRB64, err := svc.GetMDMAppleCSR(ctx)
+	if err != nil {
+		return &getMDMAppleCSRResponse{Err: err}, nil
+	}
+
+	return &getMDMAppleCSRResponse{CSR: signedCSRB64}, nil
+}
+
+func (svc *Service) GetMDMAppleCSR(ctx context.Context) (string, error) {
+	if err := svc.authz.Authorize(ctx, &fleet.AppleCSR{}, fleet.ActionWrite); err != nil {
+		return "", err
+	}
+
+	// Check if we have existing certs and keys
+	var apnsKey *rsa.PrivateKey
+	savedAssets, err := svc.ds.GetMDMConfigAssetsByName(ctx, []fleet.MDMAssetName{fleet.MDMAssetCACert, fleet.MDMAssetCAKey, fleet.MDMAssetAPNSKey})
+	if err != nil {
+		return "", ctxerr.Wrap(ctx, err, "checking asset existence")
+	}
+
+	if len(savedAssets) == 0 {
+		// Then we should create them
+		scepCert, scepKey, err := apple_mdm.NewSCEPCACertKey()
+		if err != nil {
+			return "", ctxerr.Wrap(ctx, err, "generate SCEP cert and key")
+		}
+
+		apnsKey, err = apple_mdm.NewPrivateKey()
+		if err != nil {
+			return "", ctxerr.Wrap(ctx, err, "generate new apns private key")
+		}
+
+		// Store our config assets
+		var assets []fleet.MDMConfigAsset
+		for k, v := range map[fleet.MDMAssetName][]byte{
+			fleet.MDMAssetCACert:  apple_mdm.EncodeCertPEM(scepCert),
+			fleet.MDMAssetCAKey:   apple_mdm.EncodePrivateKeyPEM(scepKey),
+			fleet.MDMAssetAPNSKey: apple_mdm.EncodePrivateKeyPEM(apnsKey),
+		} {
+			assets = append(assets, fleet.MDMConfigAsset{
+				Name:  k,
+				Value: v,
+			})
+		}
+
+		if err := svc.ds.InsertMDMConfigAssets(ctx, assets); err != nil {
+			return "", ctxerr.Wrap(ctx, err, "inserting mdm config assets")
+		}
+	} else {
+		for _, a := range savedAssets {
+			if a.Name == fleet.MDMAssetAPNSKey {
+				block, _ := pem.Decode(a.Value)
+				apnsKey, err = x509.ParsePKCS1PrivateKey(block.Bytes)
+				if err != nil {
+					return "", ctxerr.Wrap(ctx, err, "unmarshaling saved apns key")
+				}
+			}
+		}
+	}
+
+	// Generate new APNS CSR every time this is called
+	appConfig, err := svc.ds.AppConfig(ctx)
+	if err != nil {
+		return "", ctxerr.Wrap(ctx, err, "get app config")
+	}
+
+	apnsCSR, err := apple_mdm.GenerateAPNSCSR(appConfig.OrgInfo.OrgName, apnsKey)
+	if err != nil {
+		return "", ctxerr.Wrap(ctx, err, "generate APNS cert and key")
+	}
+
+	// Submit CSR to fleetdm.com for signing
+	websiteClient := fleethttp.NewClient(fleethttp.WithTimeout(10 * time.Second))
+
+	signedCSRB64, err := apple_mdm.GetSignedAPNSCSRNoEmail(websiteClient, apnsCSR)
+	if err != nil {
+		if _, ok := err.(apple_mdm.FleetWebsiteError); ok {
+			return "", ctxerr.Wrap(
+				ctx,
+				fleet.NewUserMessageError(
+					fmt.Errorf("FleetDM CSR request failed: %w", err),
+					http.StatusBadGateway,
+				),
+			)
+		}
+		return "", ctxerr.Wrap(ctx, err, "get signed CSR")
+	}
+
+	// Return signed CSR; these bytes are already base64 encoded
+	return string(signedCSRB64), nil
```

Comment:
this goes hand-to-hand withe comment I left about base64 and `[]byte`, but I think that additionally this is a bug.

- In `GetSignedAPNSCSRNoEmail`, we do `respBytes, _ := io.ReadAll(resp.Body)` and we return `respBytes`
- However, I think `respBytes` will be the full JSON response from the website, so something like `{ "csr": "..." }`
- That full `string` is getting inside `getMDMAppleCSRResponse.CSR`, so the full JSON ends up being something like:

```json
{
  "csr": "{\"csr\": \"...\"}"
}
```

---

My preference would be to unmarshal the body from the website response into a struct like:

```go
type WebsiteResponse struct {
    CSR []byte `json:"csr"`
}
```

And then marshal that raw CSR value into our struct. We could still just pass the raw json around but this is a bit safer and plays better with our framework.

I know we talked a bit about "passing along the bytes" but my understanding of that was passing along the CSR bytes without parsing them and validating that the CSR is valid

Reply:
Ah yeah you're right, I think I didn't notice this because I was returning a contrived response from my fake local server. Good catch! I like the `WebsiteResponse` approach.